File: packages/react-native-video/ios/core/VideoError.swift
//
//  VideoFileHelper.swift
//  ReactNativeVideo
//
//  Created by Krzysztof Moch on 24/01/2025.
//

import Foundation
import NitroModules

// MARK: - LibraryError
enum LibraryError: VideoError {
  case deallocated(objectName: String)
  case DRMPluginNotFound

  var code: String {
    switch self {
    case .deallocated:
      return "library/deallocated"
    case .DRMPluginNotFound:
      return "library/drm-plugin-not-found"
    }
  }

  var message: String {
    switch self {
    case let .deallocated(objectName: objectName):
      return "Object \(objectName) has been deallocated"
    case .DRMPluginNotFound:
      return "No DRM plugin have been found, please add one to the project"
    }
  }
}

// MARK: - PlayerError
enum PlayerError: VideoError {
  case notInitialized
  case assetNotInitialized
  case invalidSource
  case invalidTrackUrl(url: String)
  
  var code: String {
    switch self {
    case .notInitialized:
      return "player/not-initialized"
    case .assetNotInitialized:
      return "player/asset-not-initialized"
    case .invalidSource:
      return "player/invalid-source"
    case .invalidTrackUrl:
      return "player/invalid-track-url"
    }
  }
  
  var message: String {
    switch self {
    case .notInitialized:
      return "Player has not been initialized (Or has been set to nil)"
    case .assetNotInitialized:
      return "Asset has not been initialized (Or has been set to nil)"
    case .invalidSource:
      return "Invalid source passed to player"
    case let .invalidTrackUrl(url: url):
      return "Invalid track URL: \(url)"
    }
  }
}

// MARK: - SourceError
enum SourceError: VideoError {
  case invalidUri(uri: String)
  case missingReadFilePermission(uri: String)
  case fileDoesNotExist(uri: String)
  case failedToInitializeAsset
  case unsupportedContentType(uri: String)

  var code: String {
    switch self {
    case .invalidUri:
      return "source/invalid-uri"
    case .missingReadFilePermission:
      return "source/missing-read-file-permission"
    case .fileDoesNotExist:
      return "source/file-does-not-exist"
    case .failedToInitializeAsset:
      return "source/failed-to-initialize-asset"
    case .unsupportedContentType:
      return "source/unsupported-content-type"
    }
  }

  var message: String {
    switch self {
    case let .invalidUri(uri: uri):
      return "Invalid source file uri: \(uri)"
    case let .missingReadFilePermission(uri: uri):
      return "Missing read file permission for source file at \(uri)"
    case let .fileDoesNotExist(uri: uri):
      return "File does not exist at URI: \(uri)"
    case .failedToInitializeAsset:
      return "Failed to initialize asset"
    case let .unsupportedContentType(uri: uri):
      return "type of content (\(uri)) is not supported"
    }
  }
}

// MARK: - VideoViewError
enum VideoViewError: VideoError {
  case viewNotFound(nitroId: Double)
  case viewIsDeallocated
  case pictureInPictureNotSupported
  
  var code: String {
    switch self {
    case .viewNotFound:
      return "view/not-found"
    case .viewIsDeallocated:
      return "view/deallocated"
    case .pictureInPictureNotSupported:
      return "view/picture-in-picture-not-supported"
    }
  }
  
  var message: String {
    switch self {
    case let .viewNotFound(nitroId: nitroId):
      return "View with nitroId \(nitroId) not found"
    case .viewIsDeallocated:
      return "Attempt to access a view, but it has been deallocated (or not initialized)"
    case .pictureInPictureNotSupported:
      return "Picture in picture is not supported on this device"
    }
  }
}

// MARK: - UnknownError
struct UnknownError: VideoError {
  var code: String { "unknown/unknown" }
  var message: String { "Unknown error" }
}

// MARK: - VideoError
protocol VideoError {
  var code: String { get }
  var message: String { get }
}

extension VideoError {
  private func getMessage() -> String {
    return "{%@\(code)::\(message)@%}"
  }

  func error() -> Error {
    return RuntimeError.error(withMessage: getMessage())
  }
}

File: packages/react-native-video/android/src/main/java/com/twg/video/core/VideoError.kt
package com.twg.video.core

// Base class for all video errors
abstract class VideoError(
    open val code: String,
    message: String
) : Error("{%@$code::$message@%}")

// Library related errors
sealed class LibraryError(code: String, message: String) : VideoError(code, message) {
  object Deallocated : LibraryError(
    "library/deallocated",
    "Object has been deallocated"
  )

  object ApplicationContextNotFound : LibraryError(
    "library/application-context-not-found",
    "Application context not found"
  )

  class MethodNotSupported(val methodName: String) : LibraryError(
    "library/method-not-supported",
    "Method $methodName() is not supported on Android"
  )

  object DRMPluginNotFound : LibraryError(
    "library/drm-plugin-not-found",
    "No DRM plugin have been found, please add one to the project",
  )
}

// Player related errors
sealed class PlayerError(code: String, message: String) : VideoError(code, message) {
  object NotInitialized : PlayerError(
    "player/not-initialized",
    "Player has not been initialized (Or has been set to null)"
  )

  object AssetNotInitialized : PlayerError(
    "player/asset-not-initialized",
    "Asset has not been initialized (Or has been set to null)"
  )

  object InvalidSource : PlayerError(
    "player/invalid-source",
    "Invalid source passed to player"
  )
}

// Source related errors
sealed class SourceError(code: String, message: String) : VideoError(code, message) {
  class InvalidUri(val uri: String) : SourceError(
    "source/invalid-uri",
    "Invalid source file uri: $uri"
  )

  class MissingReadFilePermission(val uri: String) : SourceError(
    "source/missing-read-file-permission",
    "Missing read file permission for source file at $uri"
  )

  class FileDoesNotExist(val uri: String) : SourceError(
    "source/file-does-not-exist",
    "File does not exist at URI: $uri"
  )

  object FailedToInitializeAsset : SourceError(
    "source/failed-to-initialize-asset",
    "Failed to initialize asset"
  )

  class UnsupportedContentType(val uri: String) : SourceError(
    "source/unsupported-content-type",
    "type of content (${uri}) is not supported"
  )
}

// View related errors
sealed class VideoViewError(code: String, message: String) : VideoError(code, message) {
  class ViewNotFound(val viewId: Int) : VideoViewError(
    "view/not-found",
    "View with viewId $viewId not found"
  )
  object ViewIsDeallocated : VideoViewError(
    "view/deallocated",
    "Attempt to access a view, but it has been deallocated (or not initialized)"
  )
  object PictureInPictureNotSupported : VideoViewError(
    "view/picture-in-picture-not-supported",
    "Picture in picture is not supported on this device"
  )
}

// Unknown error
class UnknownError : VideoError("unknown/unknown", "Unknown error")
